lightning/src/routing/router.rs

```diff
@@ -222,67 +327,123 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 		}
 	}
 
+	// We don't want multiple paths (as per MPP) share capacity of the same channels.
+	// This map allows paths to be aware of the channel use by other paths in the same call.
+	// This would help to make a better path finding decisions and not "overbook" channels.
+	// It is currently unaware of the directions. But if we moved 1 BTC in one direction and 
+	// 1 BTC in the opposite direction, should they cancel out? Probably not, because
+	// in the worst-case order of HTLC forwarding, a channel can be overspent.
+	let mut known_channels_amounts_available_msat = HashMap::new();
+
+	// Keeping track of how much value we need to collect across channels at this iteration,
+	// so that we know which channels are definitely not worth considering: e.g., if their
+	// available capacity is only a small fraction of what we need.
+	// Also know when to stop looking for new paths.
+	let mut found_amount_msat = 0;
+
 	macro_rules! add_entry {
 		// Adds entry which goes from $src_node_id to $dest_node_id
 		// over the channel with id $chan_id with fees described in
 		// $directional_info.
-		( $chan_id: expr, $src_node_id: expr, $dest_node_id: expr, $directional_info: expr, $chan_features: expr, $starting_fee_msat: expr ) => {
+		( $chan_id: expr, $src_node_id: expr, $dest_node_id: expr, $directional_info: expr, $capacity_sats: expr, $chan_features: expr, $starting_fee_msat: expr ) => {
+			let amount_available_msat = known_channels_amounts_available_msat.entry($chan_id.clone()).or_insert_with(|| {
+				let mut amount_available_msat = None;
+				if let Some(capacity_sats) = $capacity_sats {
+					amount_available_msat = Some(capacity_sats * 1000);
+				}
+
+				if let Some(htlc_maximum_msat) = $directional_info.htlc_maximum_msat {
+					if let Some(available_msat) = amount_available_msat {
+						amount_available_msat = Some(cmp::min(available_msat, htlc_maximum_msat));
+					} else {
+						amount_available_msat = Some(htlc_maximum_msat);
+					}
+				}
+
+				match amount_available_msat {
+					Some(available_msat) => available_msat,
+					// We assume channels with unknown balance have a capacity of 0.0001 BTC (or 10_000 sats).
+					None => 10_000 * 1000
+				}
+			});
+
+			// Skip this channel if its available capacity is very low comparing to what we need at the moment.
+			let mut has_sufficient_amount = true;
+			if *amount_available_msat * 20 < recommended_available_msat - found_amount_msat {
+				has_sufficient_amount = false;
+			}
+
 			//TODO: Explore simply adding fee to hit htlc_minimum_msat
-			if $starting_fee_msat as u64 + final_value_msat >= $directional_info.htlc_minimum_msat {
-				let proportional_fee_millions = ($starting_fee_msat + final_value_msat).checked_mul($directional_info.fees.proportional_millionths as u64);
-				if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
+			let amount_to_transfer_msat: u64 = cmp::min(*amount_available_msat, recommended_available_msat);
```

Comment:
Why `recommended_available_msat` isnt that (currently) 4x our actual target?

Reply:
Yeah, this should be `final_value_msat` indeed. Updating.

Reply:
Now that I was thinking, I think it should be 4x the target, because at this point it won't hurt to collect more than we need. This amount may be spent on fees...

Reply:
This has nothing to do with stopping pathfinding btw, as that is done elsewhere below.